programs/ziti-edge-tunnel/instance-config.c

```diff
@@ -118,7 +118,23 @@ bool save_tunnel_status_to_file() {
         char* bkp_config_file_name = get_backup_config_file_name(config_path);
 
         if (mutex_initialized) {
-            uv_mutex_lock(&mutex);
+            int res = -1;
+            int try_count = 0;
+            do {
+                res = uv_mutex_trylock(&mutex);
```

Comment:
At a higher level, I'm curious why are there multiple threads that want to write this file when the tunneler is stopping? It seems inefficient at best, and I is there any concern that the first write could have different content than the second?

Assuming it makes sense to be concerned about multiple concurrent calls to this function, I think a semaphore / `uv_sem_wait` would be a better fit than a condition variable. The condition variable requires a state variable in addition to the mutex - a waiter for a condition variable is typically looking for a particular state. This function simply wants to serialize writes to the file. The indefinite wait for the semaphore should not be a problem, since no thread should have the semaphore for very long.

Reply:
When you start the tunneller, it verifies whether you still have all the json identity files in the config path. So, immediately after initiating the load function of all the identity files from the config path, the tunneller will write that information into the config path. Later while shutting down it saves the config data again, because there can be additional information related to the change in the endpoint name, the controller url etc. which would be available only after the loading is completed. The bug which I mentioned occurs only if you stop the tunneller immediately after starting it. The load operation will be running in a different thread, because it is started using "uv_work_t *loader" , that is why 2 threads are trying to save the config at the same time.

I will work on the semaphore changes that you suggested

Reply:
I changed the logic to semaphore